### ISSUE TITLE
Shell.pm not (easily) available on SCU14/15

### ISSUE THREAD

**Issue by mathomp4** [GEOS-ESM/GEOSadas#45]
Per Amal, who is trying to run ADAS on SCU14:

> I briefly tried fvsetup yesterday after I got access to SCU14/15 just to make sure everything is working. I’m getting some Shell error very early on that doesn’t seem to be something I can figure out on my own.
> ```
> Can't locate Shell.pm in @INC (you may need to install the Shell module) (@INC contains: /usr/lib/perl5/site_perl/5.18.2/x86_64-linux-thread-multi /usr/lib/perl5/site_perl/5.18.2 /usr/lib/perl5/vendor_perl/5.18.2/x86_64-linux-thread-multi /usr/lib/perl5/vendor_perl/5.18.2 /usr/lib/perl5/5.18.2/x86_64-linux-thread-multi /usr/lib/perl5/5.18.2 /usr/lib/perl5/site_perl .) at /gpfsm/dnb02/aoloso/f525_fpp_skylake/GEOSadas/Linux/bin/fvsetup line 352.
> ```

**Comment by mathomp4** [comment]
Per @aoloso and NCCS, we are pretty sure any "simple" workaround will be exciting. You can:
```bash
export PERL5LIB=/usr/lib/perl5/5.18.2/CPAN
```
but this is fragile if the version changes. You could install `Shell.pm` yourself in a non-versioned area, but that can be annoying.

You can do:
```perl
use CPAN::Shell qw(cat rm);
```
but `CPAN::Shell` does not exist on SLES11!

The real solution, I think, is to remove all dependence on `Shell.pm`.  Looking at an ADAS git clone, I see that these:
```
Applications/GEOSdas_App/g5_ddf.pl
Applications/GEOSdas_App/gen_lnbcs.pl
Applications/GEOSdas_App/setup_4merra2.pl
GMAO_Shared/GMAO_ods/obimp_summary.pl
GMAO_Shared/GMAO_etc/ana4dupd.pl
GMAO_Shared/GMAO_etc/append_gsigcrc.pl
GMAO_Shared/GMAO_etc/extract_pre-qc.pl
GMAO_Shared/GMAO_etc/verify_obsys.pl
GMAO_Shared/GMAO_etc/zeit_summary.pl
Applications/NCEP_Etc/NCEP_bias/sac.pl
Applications/NCEP_Paqc/oiqc/prepqc_daemon.pl
Applications/NCEP_Paqc/oiqc/prepqc.pl
GMAO_Shared/GEOS_Util/post/ecmwf2geos.pl
GMAO_Shared/GEOS_Util/post/ncep2geos.pl
Applications/NCEP_Etc/NCEP_enkf/scripts/gmao/jobgen.pl
Applications/NCEP_Etc/NCEP_enkf/scripts/gmao/setup_atmens.pl
GEOSgcs_GridComp/GEOSana_GridComp/GEOSaana_GridComp/GSI_GridComp/match_obcls_obsys.pl
GEOSgcs_GridComp/GEOSana_GridComp/GEOSaana_GridComp/GSI_GridComp/berror.pl
GEOSgcs_GridComp/GEOSana_GridComp/GEOSaana_GridComp/GSI_GridComp/mksi/gsiinfo.pl
```
are the Perl scripts that `use Shell`.

But, looking at each of them, I think many are just victims of copy-paste. The only ones that (I think) actually use Shell are:
```
Applications/GEOSdas_App/g5_ddf.pl
Applications/GEOSdas_App/setup_4merra2.pl
GMAO_Shared/GMAO_ods/obimp_summary.pl
GMAO_Shared/GMAO_etc/ana4dupd.pl
GMAO_Shared/GMAO_etc/zeit_summary.pl
Applications/NCEP_Paqc/oiqc/prepqc_daemon.pl
Applications/NCEP_Paqc/oiqc/prepqc.pl
GMAO_Shared/GEOS_Util/post/ecmwf2geos.pl
GMAO_Shared/GEOS_Util/post/ncep2geos.pl
```
and of these I think most are using Shell to get either 'rm' or 'cp' functionality.

Talking with @gmao-jstassi, in Perl land, `unlink()` is a built-in equivalent of `rm` and `File::Copy` is a nigh-universal module (that both SLES11 and SLES12 have) for doing file copies. 

This would be a dual repo change (here and in `GMAO_Shared`) but eventually we will be in SLES12 land at some point.

That said, I'd probably need @gmao-jstassi or @rtodling to help figure out how to test the changes. I suppose a "Does the 3d/4d ADAS" run is a good step, but are all these files used by that? Or are some used during different processes?

**Comment by rtodling** [comment]
I'll look into this since I've been planning to run on the same machines as well.

**Comment by gmao-msienkie** [comment]
These two are in my section (NCEP_Paqc) so I will have a look.  I already have a modification in OIQC pending so I can add any changes to that branch after testing.

Applications/NCEP_Paqc/oiqc/prepqc_daemon.pl
Applications/NCEP_Paqc/oiqc/prepqc.pl

(They say they use "cat" but do not, so I think I just need to test the rm->unlink.)

**Comment by mathomp4** [comment]
Yeah. I think a lot of them used to use `cat` to make a file, but got converted to here-documents. I'm sure between @gmao-jstassi and I we can figure out a more native Perl-y way to do any `Shell.pm` command. (Heck, there's always `system()`. :smiley:)

**Comment by gmao-msienkie** [comment]
maybe a few more scripts to check, including the ever popular 'analyzer' and 'diag2ods'

```
install/bin> grep -l '^use Shell qw' *
ana4dupd.pl
analyzer
append_gsigcrc.pl
diag2ods
ecmwf2geos.pl
extract_pre-qc.pl
fvarchive
fvgetf
fvsetup
g5_ddf.pl
gen_lnbcs.pl
grads2grib
gsidiags
gsiinfo.pl
jobgen.pl
match_obcls_obsys.pl
ncep2geos.pl
obimp_summary.pl
rndasfiles
rst2rcyc
rst4ncepf
sac.pl
setup_4merra2.pl
setup_atmens.pl
verify_obsys.pl
vtrack
vtxreloc
zeit_summary.pl
```

**Comment by mathomp4** [comment]
Yeah. I realized my search was only looking at the `.pl` scripts. In truth, in our work with Amal in trying to get the 4dvar on SCU14 we found that it turns out `Shell.pm` is just fundamentally *different* than it is on SLES11. To the point that the *only* way forward (in my opinion) is to remove `Shell` from the scripts entirely. I did this by hand in a checkout and...well, it let the run go further before we disbanded the meeting.

**Comment by nacks** [comment]
This appears to be a SLES11 and prior vs SLES12+. SuSE chose to move Shell.pm under the CPAN directory in their base Perl installation. You can find the CPAN Shell module via: 'use CPAN::Shell qw' rather than just "use Shell qw".

Alternatively if you don't like what SuSE has done, installing your own version of perl rather than using the system one (that we can't really change as the OS has ton of dependencies on Perl and similarly with Python).

This change appears to be in OpenSuSE as well as SLES12+

https://bugzilla.opensuse.org/show_bug.cgi?id=945657

**Comment by mathomp4** [comment]
@nacks Actually, it's worse as the `Shell.pm` on SLES12 doesn't have the same capabilities as the `Shell.pm` on SLES11. Just moving to `use CPAN::Shell` didn't work. So, the "solution" was to actually just *remove* all of Shell if I could. Most places were using it for `rm()` and `cp()` and Perl has `unlink()` and a fairly standard `File::Copy` that can do this for us.

**Comment by nacks** [comment]
Right, I was unable to get that newer CPAN::Shell to work for simple tests either. In other distributions Shell isn't even its own module. I think like you say you are better off using the built in perl5 unlink and File::Copy.

**Comment by rtodling** [comment]
I've been looking at this stuff as well - I know you've done a lot of work to the perl scripts removing references to Shell - which is good. I have to say that I caught a few cases in which your changes were not robust enough. After some fiddling w/ CPAN::Shell stuff I realize that we cannot change to use this, if nothing else, because we'll break the SLES11 build.  But I was able to go around one or two changes from you that were not general - in doing that I also found things from Joe Stassi that were not general enough and cause the build to break by simply moving the Linux (or install) directory to another location after the build had completed.
